src/librustc_trait_selection/traits/error_reporting/suggestions.rs

```diff
@@ -821,12 +822,21 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
             .iter()
             .filter_map(|expr| tables.node_type_opt(expr.hir_id))
             .map(|ty| self.resolve_vars_if_possible(&ty));
-        let (last_ty, all_returns_have_same_type) = ret_types.clone().fold(
-            (None, true),
-            |(last_ty, mut same): (std::option::Option<Ty<'_>>, bool), ty| {
+        let (last_ty, all_returns_have_same_type, never_count) = ret_types.clone().fold(
+            (None, true, 0),
+            |(last_ty, mut same, never_count): (std::option::Option<Ty<'_>>, bool, usize), ty| {
                 let ty = self.resolve_vars_if_possible(&ty);
-                same &= last_ty.map_or(true, |last_ty| last_ty == ty) && ty.kind != ty::Error;
-                (Some(ty), same)
+                same &= ty.kind != ty::Error
+                    && last_ty.map_or(true, |last_ty| {
+                        // FIXME: ideally we would use `can_coerce` here instead, but `typeck` comes
+                        // *after* in the dependency graph.
+                        match (&ty.kind, &last_ty.kind) {
+                            (Infer(InferTy::TyVar(_)), Infer(InferTy::TyVar(_))) => ty == last_ty,
+                            (Infer(a), Infer(b)) => discriminant(a) == discriminant(b),
```

Comment:
What is going on here with the `discriminant` checking?

Reply:
This is to check if two inferred literals can be coerced to the same type. For example in the following `x` and `y` will have different `ty::Infer(_)` types, but their discriminant will be the same. If we in a later compilation step we would use `can_coerce` (as mentioned in the comment), which would also make this check for us, but can't here :-/

```rust
let x = 42;
let y = 1;
if z {
    return x;
} else {
    return y;
}
```

Reply:
Does anything else do discriminant checking like this? I feel uncomfortable doing that, rather than explicitly matching.

Reply:
The first commit had the enum variants spelled out. I changed it to this [based on a review comment](https://github.com/rust-lang/rust/pull/70998#discussion_r407105709) :)

I don't see a reason _not_ to use the current code, beyond it being a bit obscure for the reader (which might be _enough_ of a reason not to use it).

Reply:
Ah, I see 😅I think the original choice was the better one.

- Discriminant checking is obscure.
- It's not clear you care about any of the cases apart from int/float.
- If `InferTy` changes in the future, it's not clear this will still be correct, but will be very hard to spot when refactoring, because it's not explicit.

Sorry for the churn!